### PR TITLE
fix(ledger): enforce block-wide execution and size limits in VerifyBlock

### LINKED ISSUE #1937: ledger: enforce block-wide execution and size limits in VerifyBlock
VerifyBlock validates each transaction and performs the Dijkstra reference-script block check, but it does not enforce the block-wide maximum execution-unit budget. The protocol parameters expose MaxBlockExUnits, MaxBlockBodySize, and MaxBlockHeaderSize, but the public block validation path does not apply the corresponding checks.

The Conway BBODY rules require total transaction exunits to remain within ppMaxExUnits, in addition to per-transaction limits. Block body and header size limits should also be enforced where VerifyBlock is intended to provide block-local validation.

Impact: a block can pass transaction-local validation while exceeding a protocol-wide resource or serialized-size limit.

References:
- https://github.com/blinklabs-io/gouroboros/blob/main/ledger/verify_block.go
- https://github.com/blinklabs-io/gouroboros/blob/main/ledger/alonzo/pparams.go
- https://cardano-scaling.github.io/cardano-blueprint/ledger/block-validation.html
- https://docs.cardano.org/about-cardano/explore-more/parameter-guide

Acceptance criteria:
- Sum transaction exunits and reject blocks above MaxBlockExUnits for applicable eras.
- Enforce applicable max block body and header sizes using the exact serialized representation required by each era.
- Add positive and negative block fixtures, including a case where every transaction is individually valid but the block total is invalid.

### PR DESCRIPTION
**Pull request by cryptodj413**
## Summary

Enforces block-wide execution unit limits and block body/header size limits in `VerifyBlock`.

- Enforces block-wide execution unit limits (sum of tx `ExUnits` vs protocol `MaxBlockExUnits`) and block body/header size limits (vs `MaxBlockBodySize`/`MaxBlockHeaderSize`) in `VerifyBlock`.
- Adds `BlockExUnitsTooBigError`, `BlockBodySizeTooBigError`, and `BlockHeaderSizeTooBigError` in `ledger/common/errors.go`, each with a sentinel var and `Is(target error) bool` method matching the existing pattern used by `RefScriptSizePerTxTooLargeError`/`RefScriptSizePerBlockTooLargeError`/`ScriptDataHashMismatchError`.
- Adds a `BlockBodySizeFromCbor` helper in `ledger/common/verify_config.go`.
- Zero-value guard: an unset (zero-value) `MaxBlockExUnits`/`MaxBlockBodySize`/`MaxBlockHeaderSize` in protocol params is treated as "no limit" and skipped, consistently across all three new checks.
- All three new block-limit errors use `ValidationErrorTypeProtocol` for consistent downstream categorization.
- New tests in `ledger/verify_block_limits_test.go` (positive case plus one failing case per limit type), plus additions to `ledger/verify_block_test.go` and `ledger/common/verify_config_test.go`.

Closes #1937

## Test plan

- `go build ./...` — clean
- `go test ./ledger/...` — passing
- `golangci-lint run ./ledger/...` — 0 issues
- `gofmt -l` on all changed files — no output (clean)

---
## Summary by cubic
Enforces block-wide execution-unit and size limits in `VerifyBlock` to match protocol rules and reject oversize blocks. Checks run before per-transaction validation; unrecognized `ProtocolParameters` now fail closed with a configuration error.

- **Bug Fixes**
  - Enforce block-wide `ExUnits` vs `MaxBlockExUnits`, and header/body sizes vs `MaxBlockHeaderSize`/`MaxBlockBodySize` using preserved CBOR (`header.Cbor()`; body via `BlockBodySizeFromCbor`, only when the limit > 0).
  - Add `SkipBlockLimitsValidation` in `VerifyConfig`; independent of transaction validation and does not require `LedgerState`.
  - Treat zero-value limits as “no limit”; unrecognized `ProtocolParameters` return `ValidationErrorTypeConfiguration` (explicit opt-out is `SkipBlockLimitsValidation`).
  - Reject negative `ExUnits.Memory/Steps` and detect overflow; introduce `BlockExUnitsTooBigError`, `BlockBodySizeTooBigError`, `BlockHeaderSizeTooBigError` (all `ValidationErrorTypeProtocol`), with tests for positive/negative paths and overflow.

<sup>Written for commit ee5bb8ea83aefab35b8911d0f8e9552ead862388. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/blinklabs-io/gouroboros/pull/1950?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added block-wide validation for execution-unit totals, serialized body size, and header size against protocol limits.
  * Added clear, typed errors when blocks exceed configured limits.
  * Added support for calculating block body size from CBOR data.
  * Added configuration to skip block-limit validation when needed.

* **Bug Fixes**
  * Improved detection of malformed or incomplete block data and execution-unit total overflow.

* **Tests**
  * Added coverage for valid blocks, exceeded limits, multiple eras, and malformed CBOR.

### CODE REVIEW

File: ledger/verify_block.go [hunk 1]
```diff
@@ -674,6 +776,135 @@ func VerifyBlock(
...
+					common.ValidationErrorTypeConfiguration,
+					"unable to determine block-wide limits from protocol parameters",
+					map[string]any{
+						"block_slot":   slot,
+						"block_number": blockNo,
+						"era":          era,
+					},
+					limitsErr,
+				)
+			}
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply ledger/verify_block.go:783]: Fixed in 56c0d8c: `blockLevelLimits`'s `default` case for an unrecognized `ProtocolParameters` implementation (e.g. `mockledger.MockProtocolParamsRules`) no longer returns a configuration error. It now returns `(0, 0, common.ExUnits{}, false)` — i.e. "no limits configured" — so block-limit checks are silently skipped for that caller instead of rejecting the block. `SkipBlockLimitsValidation` remains the explicit opt-out for callers with a supported pparams type who still want to disable the check. See the updated doc comment on `blockLevelLimits` in `ledger/verify_block.go`.

File: ledger/verify_block.go [hunk 2]
```diff
@@ -674,6 +776,135 @@ func VerifyBlock(
...
+							"era":             era,
+							"header_size":     headerSize,
+							"max_header_size": maxHeaderSize,
+						},
+						common.BlockHeaderSizeTooBigError{
+							HeaderSize:         headerSize,
+							MaxBlockHeaderSize: maxHeaderSize,
+						},
+					)
+				}
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply ledger/verify_block.go]: Fixed in 56c0d8c: applied essentially the restructure you proposed. The header-size check is now computed from `headerCbor := header.Cbor()` independently, gated on `len(headerCbor) > 0 && maxHeaderSize > 0`, and runs before/outside the `len(rawBlockCbor) > 0` guard that the body-size check uses. So a block with header CBOR available but no full raw block CBOR still gets `ppMaxBlockHeaderSize` enforced.

File: ledger/verify_block.go [hunk 3]
```diff
@@ -308,6 +308,108 @@ func extractHeaderFields(
...
+			nil
+	case *dijkstra.DijkstraProtocolParameters:
+		return uint64(p.MaxBlockBodySize),
+			uint64(p.MaxBlockHeaderSize),
+			p.MaxBlockExUnits,
+			true,
+			nil
+	default:
+		return 0, 0, common.ExUnits{}, false, fmt.Errorf(
+			"unsupported protocol parameters type %T for block-limit validation",
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply ledger/verify_block.go:377]: Addressed in 56c0d8c (pushed to this branch):

1. `blockLevelLimits`'s `default` case no longer fails closed on an unrecognized `ProtocolParameters` implementation — it's now treated as "no limits configured" and block-limit checks are skipped for that caller, so callers passing e.g. `mockledger.MockProtocolParamsRules` aren't broken. `SkipBlockLimitsValidation` remains the explicit opt-out. (verify_block.go:366)
2. Block-wide limit checks (ExUnits budget + body/header size) now run before the per-transaction validation loop in `VerifyBlock`, so an oversized/over-budget block is rejected before paying the cost of validating every transaction/Plutus script in it. (verify_block.go:795)
3. The header-size check no longer depends on `len(rawBlockCbor) > 0` — it's derived from `header.Cbor()` independently and checked whenever header CBOR is present. (verify_block.go:856)
4. Body-size calculation (`BlockBodySizeFromCbor`) is now skipped entirely when `MaxBlockBodySize == 0`, rather than always computed with only the comparison skipped. (verify_block.go:874)
5. On the `BlockBodySizeFromCbor` full-CBOR-redecode perf concern (verify_config.go:205): **not fully fixed in this PR** — added a `NOTE:` doc comment documenting it as a known perf follow-up requiring a larger refactor (deriving body size without re-tokenizing the whole block, rather than the current full redecode). Fix #4 above (skipping the call entirely when the limit is disabled) is the low-risk mitigation applied now; a full fix touches the decode pipeline more broadly than this PR's scope.
6. Added the missing zero-limit "unlimited" test coverage you flagged — `TestVerifyBlock_BlockLimits_ZeroLimitsAreUnlimited` in `ledger/verify_block_limits_test.go`, asserting a block passes when `MaxBlockExUnits`/`MaxBlockBodySize`/`MaxBlockHeaderSize` are all zero. (verify_block_limits_test.go:125)

File: ledger/verify_block.go [hunk 4]
```diff
@@ -674,6 +776,135 @@ func VerifyBlock(
...
+	// a block-local structural/resource check, not a per-transaction UTxO
+	// rule, so it needs neither LedgerState nor a full per-transaction pass.
+	// It only runs when there is something to check (at least one
+	// transaction, for the ExUnits budget, or the block's raw CBOR is
+	// available, for the size checks) and ProtocolParameters is set;
+	// blockLevelLimits type-asserts config.ProtocolParameters to the era's
+	// concrete pparams struct (there is no shared getter across eras, and
+	// no other way to confirm it matches the block's actual era), mirroring
+	// the same assumption the per-era rules make.
+	if block.Era() != byron.EraByron && !config.SkipBlockLimitsValidation &&
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply ledger/verify_block.go:763]: Addressed in 56c0d8c (pushed to this branch):

1. `blockLevelLimits`'s `default` case no longer fails closed on an unrecognized `ProtocolParameters` implementation — it's now treated as "no limits configured" and block-limit checks are skipped for that caller, so callers passing e.g. `mockledger.MockProtocolParamsRules` aren't broken. `SkipBlockLimitsValidation` remains the explicit opt-out. (verify_block.go:366)
2. Block-wide limit checks (ExUnits budget + body/header size) now run before the per-transaction validation loop in `VerifyBlock`, so an oversized/over-budget block is rejected before paying the cost of validating every transaction/Plutus script in it. (verify_block.go:795)
3. The header-size check no longer depends on `len(rawBlockCbor) > 0` — it's derived from `header.Cbor()` independently and checked whenever header CBOR is present. (verify_block.go:856)
4. Body-size calculation (`BlockBodySizeFromCbor`) is now skipped entirely when `MaxBlockBodySize == 0`, rather than always computed with only the comparison skipped. (verify_block.go:874)
5. On the `BlockBodySizeFromCbor` full-CBOR-redecode perf concern (verify_config.go:205): **not fully fixed in this PR** — added a `NOTE:` doc comment documenting it as a known perf follow-up requiring a larger refactor (deriving body size without re-tokenizing the whole block, rather than the current full redecode). Fix #4 above (skipping the call entirely when the limit is disabled) is the low-risk mitigation applied now; a full fix touches the decode pipeline more broadly than this PR's scope.
6. Added the missing zero-limit "unlimited" test coverage you flagged — `TestVerifyBlock_BlockLimits_ZeroLimitsAreUnlimited` in `ledger/verify_block_limits_test.go`, asserting a block passes when `MaxBlockExUnits`/`MaxBlockBodySize`/`MaxBlockHeaderSize` are all zero. (verify_block_limits_test.go:125)

File: ledger/verify_block.go [hunk 5]
```diff
@@ -674,6 +776,135 @@ func VerifyBlock(
...
+						},
+						common.BlockExUnitsTooBigError{
+							TotalExUnits:    totalExUnits,
+							MaxBlockExUnits: maxExUnits,
+						},
+					)
+				}
+			}
+			if len(rawBlockCbor) > 0 {
+				headerSize := uint64(len(header.Cbor()))
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply ledger/verify_block.go]: Addressed in 56c0d8c (pushed to this branch):

1. `blockLevelLimits`'s `default` case no longer fails closed on an unrecognized `ProtocolParameters` implementation — it's now treated as "no limits configured" and block-limit checks are skipped for that caller, so callers passing e.g. `mockledger.MockProtocolParamsRules` aren't broken. `SkipBlockLimitsValidation` remains the explicit opt-out. (verify_block.go:366)
2. Block-wide limit checks (ExUnits budget + body/header size) now run before the per-transaction validation loop in `VerifyBlock`, so an oversized/over-budget block is rejected before paying the cost of validating every transaction/Plutus script in it. (verify_block.go:795)
3. The header-size check no longer depends on `len(rawBlockCbor) > 0` — it's derived from `header.Cbor()` independently and checked whenever header CBOR is present. (verify_block.go:856)
4. Body-size calculation (`BlockBodySizeFromCbor`) is now skipped entirely when `MaxBlockBodySize == 0`, rather than always computed with only the comparison skipped. (verify_block.go:874)
5. On the `BlockBodySizeFromCbor` full-CBOR-redecode perf concern (verify_config.go:205): **not fully fixed in this PR** — added a `NOTE:` doc comment documenting it as a known perf follow-up requiring a larger refactor (deriving body size without re-tokenizing the whole block, rather than the current full redecode). Fix #4 above (skipping the call entirely when the limit is disabled) is the low-risk mitigation applied now; a full fix touches the decode pipeline more broadly than this PR's scope.
6. Added the missing zero-limit "unlimited" test coverage you flagged — `TestVerifyBlock_BlockLimits_ZeroLimitsAreUnlimited` in `ledger/verify_block_limits_test.go`, asserting a block passes when `MaxBlockExUnits`/`MaxBlockBodySize`/`MaxBlockHeaderSize` are all zero. (verify_block_limits_test.go:125)

File: ledger/verify_block.go [hunk 6]
```diff
@@ -674,6 +776,135 @@ func VerifyBlock(
...
+							"header_size":     headerSize,
+							"max_header_size": maxHeaderSize,
+						},
+						common.BlockHeaderSizeTooBigError{
+							HeaderSize:         headerSize,
+							MaxBlockHeaderSize: maxHeaderSize,
+						},
+					)
+				}
+				bodySize, sizeErr := common.BlockBodySizeFromCbor(rawBlockCbor)
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply ledger/verify_block.go:854]: Addressed in 56c0d8c (pushed to this branch):

1. `blockLevelLimits`'s `default` case no longer fails closed on an unrecognized `ProtocolParameters` implementation — it's now treated as "no limits configured" and block-limit checks are skipped for that caller, so callers passing e.g. `mockledger.MockProtocolParamsRules` aren't broken. `SkipBlockLimitsValidation` remains the explicit opt-out. (verify_block.go:366)
2. Block-wide limit checks (ExUnits budget + body/header size) now run before the per-transaction validation loop in `VerifyBlock`, so an oversized/over-budget block is rejected before paying the cost of validating every transaction/Plutus script in it. (verify_block.go:795)
3. The header-size check no longer depends on `len(rawBlockCbor) > 0` — it's derived from `header.Cbor()` independently and checked whenever header CBOR is present. (verify_block.go:856)
4. Body-size calculation (`BlockBodySizeFromCbor`) is now skipped entirely when `MaxBlockBodySize == 0`, rather than always computed with only the comparison skipped. (verify_block.go:874)
5. On the `BlockBodySizeFromCbor` full-CBOR-redecode perf concern (verify_config.go:205): **not fully fixed in this PR** — added a `NOTE:` doc comment documenting it as a known perf follow-up requiring a larger refactor (deriving body size without re-tokenizing the whole block, rather than the current full redecode). Fix #4 above (skipping the call entirely when the limit is disabled) is the low-risk mitigation applied now; a full fix touches the decode pipeline more broadly than this PR's scope.
6. Added the missing zero-limit "unlimited" test coverage you flagged — `TestVerifyBlock_BlockLimits_ZeroLimitsAreUnlimited` in `ledger/verify_block_limits_test.go`, asserting a block passes when `MaxBlockExUnits`/`MaxBlockBodySize`/`MaxBlockHeaderSize` are all zero. (verify_block_limits_test.go:125)

File: ledger/common/verify_config.go [hunk 7]
```diff
@@ -175,3 +189,35 @@ func ValidateBlockBodyHash(
...
+//
+// This matches cardano-ledger's default blockBodySize, which for
+// pre-Dijkstra blocks is the concatenation of transaction_bodies,
+// transaction_witness_sets, auxiliary_data_set, and (Alonzo+)
+// invalid_transactions, each encoded as normal top-level CBOR values with no
+// extra wrapping array around the group; and for Dijkstra (a 2-element
+// [header, block_body] block) is simply the single block_body element.
+// Summing raw[1:] covers both shapes without era-specific branching.
+func BlockBodySizeFromCbor(rawCbor []byte) (uint64, error) {
+	var raw []cbor.RawMessage
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply ledger/common/verify_config.go:218]: Addressed in 56c0d8c (pushed to this branch):

1. `blockLevelLimits`'s `default` case no longer fails closed on an unrecognized `ProtocolParameters` implementation — it's now treated as "no limits configured" and block-limit checks are skipped for that caller, so callers passing e.g. `mockledger.MockProtocolParamsRules` aren't broken. `SkipBlockLimitsValidation` remains the explicit opt-out. (verify_block.go:366)
2. Block-wide limit checks (ExUnits budget + body/header size) now run before the per-transaction validation loop in `VerifyBlock`, so an oversized/over-budget block is rejected before paying the cost of validating every transaction/Plutus script in it. (verify_block.go:795)
3. The header-size check no longer depends on `len(rawBlockCbor) > 0` — it's derived from `header.Cbor()` independently and checked whenever header CBOR is present. (verify_block.go:856)
4. Body-size calculation (`BlockBodySizeFromCbor`) is now skipped entirely when `MaxBlockBodySize == 0`, rather than always computed with only the comparison skipped. (verify_block.go:874)
5. On the `BlockBodySizeFromCbor` full-CBOR-redecode perf concern (verify_config.go:205): **not fully fixed in this PR** — added a `NOTE:` doc comment documenting it as a known perf follow-up requiring a larger refactor (deriving body size without re-tokenizing the whole block, rather than the current full redecode). Fix #4 above (skipping the call entirely when the limit is disabled) is the low-risk mitigation applied now; a full fix touches the decode pipeline more broadly than this PR's scope.
6. Added the missing zero-limit "unlimited" test coverage you flagged — `TestVerifyBlock_BlockLimits_ZeroLimitsAreUnlimited` in `ledger/verify_block_limits_test.go`, asserting a block passes when `MaxBlockExUnits`/`MaxBlockBodySize`/`MaxBlockHeaderSize` are all zero. (verify_block_limits_test.go:125)

File: ledger/verify_block_limits_test.go [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+		SkipTransactionValidation: true,
+		SkipStakePoolValidation:   true,
+		ProtocolParameters:        pp,
+	}
+}
+
+// TestVerifyBlock_BlockLimits_Positive is the positive fixture: a block
+// whose total ExUnits, body size, and header size are all comfortably
+// within protocol maxima passes VerifyBlock.
+func TestVerifyBlock_BlockLimits_Positive(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply ledger/verify_block_limits_test.go:125]: Addressed in 56c0d8c (pushed to this branch):

1. `blockLevelLimits`'s `default` case no longer fails closed on an unrecognized `ProtocolParameters` implementation — it's now treated as "no limits configured" and block-limit checks are skipped for that caller, so callers passing e.g. `mockledger.MockProtocolParamsRules` aren't broken. `SkipBlockLimitsValidation` remains the explicit opt-out. (verify_block.go:366)
2. Block-wide limit checks (ExUnits budget + body/header size) now run before the per-transaction validation loop in `VerifyBlock`, so an oversized/over-budget block is rejected before paying the cost of validating every transaction/Plutus script in it. (verify_block.go:795)
3. The header-size check no longer depends on `len(rawBlockCbor) > 0` — it's derived from `header.Cbor()` independently and checked whenever header CBOR is present. (verify_block.go:856)
4. Body-size calculation (`BlockBodySizeFromCbor`) is now skipped entirely when `MaxBlockBodySize == 0`, rather than always computed with only the comparison skipped. (verify_block.go:874)
5. On the `BlockBodySizeFromCbor` full-CBOR-redecode perf concern (verify_config.go:205): **not fully fixed in this PR** — added a `NOTE:` doc comment documenting it as a known perf follow-up requiring a larger refactor (deriving body size without re-tokenizing the whole block, rather than the current full redecode). Fix #4 above (skipping the call entirely when the limit is disabled) is the low-risk mitigation applied now; a full fix touches the decode pipeline more broadly than this PR's scope.
6. Added the missing zero-limit "unlimited" test coverage you flagged — `TestVerifyBlock_BlockLimits_ZeroLimitsAreUnlimited` in `ledger/verify_block_limits_test.go`, asserting a block passes when `MaxBlockExUnits`/`MaxBlockBodySize`/`MaxBlockHeaderSize` are all zero. (verify_block_limits_test.go:125)

**Comment by cryptodj413** [comment]
Addressing CodeRabbit's two nitpick comments from the review (embedded in the review summary, not separate inline threads, so replying here instead):

- **`ledger/verify_block_test.go` (939-969): "Add a subtest for the overflow branch."** Added — a new subtest covering `sumBlockExUnits` returning an error when the summed `Memory` ExUnits overflow `math.MaxInt64`.
- **`ledger/verify_block_limits_test.go` (122-148): "Add a zero-limit case."** Added — `TestVerifyBlock_BlockLimits_ZeroLimitsAreUnlimited`, which sets `MaxBlockExUnits`, `MaxBlockBodySize`, and `MaxBlockHeaderSize` all to zero and asserts the block still passes `VerifyBlock`.

Both are in 56c0d8c, along with the fixes for the other CodeRabbit/cubic-dev-ai findings on this PR (replied individually on each of those inline threads).

File: ledger/verify_block.go [hunk 9]
```diff
@@ -308,6 +308,114 @@ func extractHeaderFields(
...
+			uint64(p.MaxBlockHeaderSize),
+			p.MaxBlockExUnits,
+			true
+	default:
+		// Unsupported/unrecognized ProtocolParameters implementation: treat
+		// as "no limits configured" rather than failing the whole block.
+		// SkipBlockLimitsValidation remains the explicit way to opt out;
+		// this is the implicit fallback for types that simply don't carry
+		// block-limit fields.
+		return 0, 0, common.ExUnits{}, false
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply ledger/verify_block.go]: Fixed in 24410fe: added `slog.Warn("unrecognized ProtocolParameters implementation: block-wide size and ExUnits limits are not enforced", "protocol_parameters_type", fmt.Sprintf("%T", pp))` in the `default:` branch of `blockLevelLimits`, matching the existing `slog.Warn` convention already used in `ledger/common/pparams.go:65` for a similar unrecognized-value fail-open case. Fail-open behavior (zero limits returned, no error) is unchanged — this is purely diagnostic. Extended `TestBlockLevelLimits`'s "unrecognized type is treated as no limits configured" subtest in `ledger/verify_block_test.go` to assert the WARN log line is emitted, in addition to the existing fail-open assertions.

File: ledger/verify_block.go [hunk 10]
```diff
@@ -308,6 +309,122 @@ func extractHeaderFields(
...
+		// as "no limits configured" rather than failing the whole block.
+		// SkipBlockLimitsValidation remains the explicit way to opt out;
+		// this is the implicit fallback for types that simply don't carry
+		// block-limit fields. Surface this with a warning (matching the
+		// slog.Warn convention already used in
+		// ledger/common/pparams.go:ConvertToUtxorpcCardanoCostModels) so an
+		// unexpected pparams type isn't a fully silent no-op, without
+		// changing the fail-open behavior itself.
+		slog.Warn(
+			"unrecognized ProtocolParameters implementation: block-wide size and ExUnits limits are not enforced",
```
Comment (wolf31o2) [inline ledger/verify_block.go]: This fail-open fallback disables every block-wide limit when the caller supplies an unrecognized ProtocolParameters implementation. VerifyBlock then accepts blocks exceeding ppMaxBlockBodySize, ppMaxBlockHeaderSize, or ppMaxBlockExUnits even though SkipBlockLimitsValidation is false. Please return a configuration error here (or require an explicit opt-out) rather than treating an unknown parameter type as no limits.

File: ledger/verify_block.go [hunk 11]
```diff
@@ -308,6 +309,122 @@ func extractHeaderFields(
...
+	for _, tx := range txs {
+		witnesses := tx.Witnesses()
+		if witnesses == nil {
+			continue
+		}
+		redeemers := witnesses.Redeemers()
+		if redeemers == nil {
+			continue
+		}
+		for _, value := range redeemers.Iter() {
```
Comment (wolf31o2) [inline ledger/verify_block.go:399]: ExUnits are represented as signed int64, but protocol execution units are non-negative. A negative Memory or Steps value is currently included in the sum without validation, so malformed redeemers can reduce the block total and bypass the maximum. Reject negative values before adding them (and add a regression test using decoded negative ExUnits).

**Review by wolf31o2 — CHANGES_REQUESTED** [review]
Request changes: the block-limit checks can be bypassed in two ways described by the inline comments below. Please address both before merge.

File: ledger/verify_block.go (hunk 10 above)
Comment (cryptodj413) [reply ledger/verify_block.go]: Reverted this in ee5bb8e: `blockLevelLimits`'s `default` case now returns a configuration error (`fmt.Errorf("unsupported protocol parameters type %T for block-limit validation", pp)`) instead of falling back to zero limits. The `VerifyBlock` call site surfaces this as a hard rejection via `common.NewValidationError(common.ValidationErrorTypeConfiguration, ...)`.

Since `SkipBlockLimitsValidation` already gates the entire block-limits section before `blockLevelLimits` is ever invoked, it is now the sole sanctioned explicit opt-out, exactly as you suggested ("return a configuration error here (or require an explicit opt-out)"). Callers who legitimately don't want block-limit enforcement (e.g. because they pass a mock `ProtocolParameters` unrelated to block limits) must set `SkipBlockLimitsValidation: true` rather than relying on an unrecognized type to silently disable the check.

Added `TestVerifyBlock_UnrecognizedProtocolParameters` (end-to-end through `VerifyBlock`, not just `blockLevelLimits` in isolation) with two subtests: one proving the fail-closed configuration error, and one proving the `SkipBlockLimitsValidation: true` opt-out still works. `TestBlockLevelLimits`'s existing subtest was restored to assert the hard error again.

File: ledger/verify_block.go (hunk 11 above)
Comment (cryptodj413) [reply ledger/verify_block.go:399]: Fixed in ee5bb8e: `sumBlockExUnits` now checks each redeemer's `ExUnits.Memory` and `ExUnits.Steps` before folding it into the running total via `common.AddInt64Checked`, and returns an error (`fmt.Errorf("negative execution-unit memory/steps in redeemer: %d", ...)`) if either is negative, instead of silently summing it.

Added two regression tests in `TestSumBlockExUnits` ("negative decoded memory is rejected" and "negative decoded steps is rejected") that decode real negative CBOR integers (major type 1, e.g. `0x82 0x20 0x05` -> `[-1, 5]`) into `common.ExUnits`, confirm the decoded value is actually negative, then assert `sumBlockExUnits` rejects it rather than including it in the block-wide total.

**Review by wolf31o2 — APPROVED** [review]
